runtime/include/chpl-qsbr.h

```diff
@@ -0,0 +1,169 @@
+#include "chpl-env.h"
+
+#ifndef _chpl_qsbr_h_
+#define _chpl_qsbr_h_
+
+/*
+ Concurrent-safe memory reclamation using Quiescent State-Based Reclamation (QSBR).
+
+ For memory to be appropriately protected by QSBR, two invariants must be established:
+  1.  References to QSBR-protected memory obtained prior to a checkpoint must not be
+      used after the checkpoint, otherwise the reference is unsafe.
+  2.  QSBR-protected memory must only be deferred for deletion at most once.
+
+ The idiomatic usage of QSBR involves separating memory reclamation into two phases:
+  1.  Logically deleting memory by removing it from some data structure in some linearizable fashion.
+  2.  Physically deleting the memory by deferring deletion of it.
+
+ This is commonly the case for non-blocking data structures and algorithms. For an example, see below:
+
+ ======================================
+      Non-Blocking Stack
+ ======================================
+  struct node *head;
+
+  void *dequeue(void) {
+    struct node *currentHead, *newHead;
+    // The currentHead is logically removed from the list...
+    do {
+      currentHead = load(&head);
+      if (currentHead == NULL) return NULL;
+      newHead = currentHead->next;
+    } while (CAS(&head, currentHead, newHead));
+    
+    // Queue up for physical removal...
+    void *ret = currentHead->value;
+    chpl_qsbr_defer_delete(currentHead);
+    return ret;
+  }
+*/
+
+static const int CHPL_QSBR_ITERATIONS_PER_CHECKPOINT = 64;
+
+// Thread-specific meta data.
+struct tls_node {
+  /*
+    The current observed epoch. This is used to ensure that writers
+    do not delete an older instance that corresponds to this epoch.
+  */
+  atomic_uint_least64_t epoch;
+  /*
+    Whether or not this thread is parked in the runtime. If the thread is
+    parked, it is ignored during memory reclamation checks.
+  */
+  atomic_uint_least64_t parked;
+  /*
+    List of objects that are queued for deferred deletion, sorted by target_epoch. 
+    We defer deletion if we find that while scanning the list of all thread-specific
+    data, they are not up-to-date on that epoch. By sorting by target_epoch, we can
+    make shortcuts by removing a large bulk of data at once. The list is protected
+    by a lightweight spinlock, but is acquired using pseudo-TestAndSet operations
+    due to low contention.
+  */
+  atomic_uint_least64_t spinlock;
+  struct defer_node *deferList;
+
+  /*
+    Recycled defer nodes that can be reused. This is not a thread-safe field and
+    only be accessed by the thread that owns this node.
+  */
+  struct defer_node *recycleList;
+
+  /*
+    Next TLS node. Note that this field can be updated atomically by newer threads
+    and so should be accessed via atomics.
+  */
+  struct tls_node *next;
+};
+
+extern CHPL_TLS_DECL(struct tls_node *, chpl_qsbr_tls);
+
+#define chpl_qsbr_likely(x)       __builtin_expect((x),1)
+#define chpl_qsbr_unlikely(x)     __builtin_expect((x),0)
```

Comment:
I think you need to wrap this in an ifdef, since some C compilers won't support __builtin_expect.

See
runtime/include/chpl-comp-detect-macros.h

Reply:
Since its available on clang, gcc, and craycc, is there an official way to determine if the feature is supported? I believe it should be added correctly in its own header or some other widely-available header as I still believe it can come in handy.

Reply:
Perhaps moving it to its own header and using ifdefs to enable it on those 3 you know it works for is good enough for now.

Reply:
I'll just remove them for now.